src/java/org/apache/cassandra/transport/messages/ExecuteMessage.java

```diff
@@ -128,38 +137,55 @@ protected Message.Response execute(QueryState state, long queryStartNanoTime, bo
 
             if (!prepared.fullyQualified
-                && !Objects.equals(state.getClientState().getRawKeyspace(), prepared.keyspace)
+                && !Objects.equals(queryState.getClientState().getRawKeyspace(), prepared.keyspace)
                 // We can not reliably detect inconsistencies for batches yet
-                && !(prepared.statement instanceof BatchStatement)
-            )
+                && !(prepared.statement instanceof BatchStatement))
             {
-                state.getClientState().warnAboutUseWithPreparedStatements(statementId, prepared.keyspace);
+                queryState.getClientState().warnAboutUseWithPreparedStatements(statementId, prepared.keyspace);
                 String msg = String.format("Tried to execute a prepared unqalified statement on a keyspace it was not prepared on. " +
                                            "Executing the resulting prepared statement will return unexpected results: %s (on keyspace %s, previously prepared on %s)",
-                                           statementId, state.getClientState().getRawKeyspace(), prepared.keyspace);
+                                           statementId, queryState.getClientState().getRawKeyspace(), prepared.keyspace);
                 nospam.error(msg);
             }
 
-
             CQLStatement statement = prepared.statement;
             options.prepare(statement.getBindVariables());
 
             if (options.getPageSize().getSize() == 0)
                 throw new ProtocolException("The page size cannot be 0");
 
             if (traceRequest)
-                traceQuery(state, prepared);
+                traceQuery(queryState, prepared);
 
             // Some custom QueryHandlers are interested by the bound names. We provide them this information
             // by wrapping the QueryOptions.
             QueryOptions queryOptions = QueryOptions.addColumnSpecifications(options, prepared.statement.getBindVariables());
 
-            long requestStartTime = System.currentTimeMillis();
+            long requestStartMillisTime = System.currentTimeMillis();
+            Optional<Stage> asyncStage = Stage.fromStatement(statement);
+            if (asyncStage.isPresent())
+            {
+                QueryHandler.Prepared finalPrepared = prepared;
+                return CompletableFuture.supplyAsync(() -> handleRequest(queryState, queryStartNanoTime, handler, queryOptions, statement, finalPrepared, requestStartMillisTime),
+                                                     asyncStage.get().executor());
+            }
+            else
+                return CompletableFuture.completedFuture(handleRequest(queryState, queryStartNanoTime, handler, queryOptions, statement, prepared, requestStartMillisTime));
+        }
+        catch (Exception e)
+        {
+            return CompletableFuture.completedFuture(handleException(queryState, prepared, e));
+        }
```

Comment:
this pattern is repeated a couple of times.
While I don't insist, it might be worthwile to extract is as a function perhaps?

Reply:
I'd like to keep it like that to make the exception handling explicit.